core/include/stdio.h

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (C) 2024 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ */
+
+/**
+ * @ingroup     sys_stdio_null
+ *
+ * This module a wrapper for the stdio.h header intended to remove all calls
+ * to stdout when stdio_null is used.
+ *
+ * This needs to reside in `core/` so it gets included early.
+ *
+ * @{
+ *
+ * @file
+ * @brief       stdio wrapper to replace the C libs stdio
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ */
+
+#ifndef CORE_STDIO_H
+#define CORE_STDIO_H
+#include_next "stdio.h"
```

Comment:
That's a GNU extension, AIU supported by Clang as well.

I'm not saying we shouldn't do this (I'm glad of all libc headers we can do ourselves to avoid weird choices by the used libc), but it is in direct conflict with the first bulled of the coding convention. But if we start doing it, better change the code convention (or if that's what we do get a 2-ACK dispensation), and reap the fruit by using `#pragma once`, rather than creating a divide between what we do and what we claim to do.

Reply:
+1 for updating the coding convention to match.

We actually already have an used of `#include_next` in the AVR code to automatically move format string to flash and use `printf_P()` to save lots of RAM.

Reply:
I added a line to `CODING_CONVENTIONS.md`.
I think being implemented by both GCC and Clang is enough of a safeguard to prevent opening the gates to [some crazy things](https://gcc.gnu.org/onlinedocs/gcc/Nested-Functions.html).

Reply:
We already use `#include_next` in `esp32`, `tinyusb`, `posix` and `avr8_common` - is the `stdio` case really special?

Reply:
AVR can use upstream GCC (and we do so). Upstream clang is rather mature, but not feature complete.

ESP does indeed need magic binary only compilers.